image-mdraid.c

```diff
@@ -0,0 +1,226 @@
+	sb->cnt_corrected_read=0; /* number of read errors that were corrected by re-writing */
+	random_uuid(sb->device_uuid); /* user-space setable, ignored by kernel U8[16] */ //TODO: should we allow user to set this?
+	sb->devflags=0;	/* per-device flags.  Only two defined...*/
+		//#define	WriteMostly1	1	/* mask for writemostly flag in above */
+		//#define	FailFast1	2	/* Should avoid retries and fixups and just fail */
+
+		/* Bad block log.  If there are any bad blocks the feature flag is set.
+		* If offset and size are non-zero, that space is reserved and available
+		*/
+	sb->bblog_shift=9;    /* shift from sectors to block size */ //TODO: not sure why this is 9
+	sb->bblog_size=8; /* number of sectors reserved for list */
+	sb->bblog_offset=16;   /* sector offset from superblock to bblog,
+			* signed - not unsigned */
+
+	/* array state information - 64 bytes */
+	sb->utime=0;		/* 40 bits second, 24 bits microseconds */
+	sb->events=0;		/* incremented when superblock updated */
+	sb->resync_offset=0;	/* data before this offset (from data_offset) known to be in sync */
+	sb->max_dev=max_devices; /* size of devs[] array to consider */
+	//__u8	pad3[64-32];	/* set to 0 when writing */
+
+	/* device state information. Indexed by dev_number.
+	 * 2 bytes per device
+	 * Note there are no per-device state flags. State information is rolled
+	 * into the 'roles' value.  If a device is spare or faulty, then it doesn't
+	 * have a meaningful role.
+	 */
+	//__le16	dev_roles[];	/* role in array, or 0xffff for a spare, or 0xfffe for faulty */
+
+
+	//Calculate checksum
+	sb->sb_csum=calc_sb_1_csum(sb);
+
+
+	//construct image file
+	int ret;
+	ret = prepare_image(image, image->size);
+	if(ret) return ret;
+	ret = insert_data(image, sb, imageoutfile(image), sizeof(struct mdp_superblock_1)+max_devices*2, 8*512);
+	if(ret) return ret;
+	if(img_in) {
+		ret = insert_image(image, img_in, image->size-DATA_OFFSET_BYTES, DATA_OFFSET_BYTES, 0);
+		if(ret) return ret;
+	}
+
+	free(sb);
+
+	return 0;
+}
+
+static int mdraid_setup(struct image *image, cfg_t *cfg) {
+	srand(time(NULL)); //TODO: Should we seed UUID in more unique way?
+
+	int raid_level = cfg_getint(image->imagesec, "level");
+
+	if(raid_level != 1) {
+		image_error(image, "MDRAID Currently only supporting raid level 1 (mirror)!\n");
+		return 1;
+	}
+
+
+	struct image *img_in = NULL;
+	struct partition *part;
+	list_for_each_entry(part, &image->partitions, list) {
+		if(strcmp(part->name, "data")) {
+			image_info(image, "MDRAID partition has to be called 'data' instead of '%s'\n", part->name);
+		} else {
+			if(img_in) {
+				image_error(image, "MDRAID cannot contain more than one data partition!\n");
+				return 2;
+			}
+			if(part->image) {
+				image_info(image, "MDRAID using data from [%s]: %s\n", part->name, part->image);
+				img_in = image_get(part->image);
+				if(image->size < (img_in->size + DATA_OFFSET_BYTES)) image->size = (img_in->size + DATA_OFFSET_BYTES);
```

Comment:
No. Don't overwrite the size silently. Set it if the size is zero. If it's too small, fail with an error.

And put the assignment in a new line.